<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Gene Set Enrichment Analysis with ggpicrust2</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Gene Set Enrichment Analysis with
ggpicrust2</h1>



<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>This vignette demonstrates how to perform Gene Set Enrichment
Analysis (GSEA) on PICRUSt2 predicted functional data using the
ggpicrust2 package. GSEA is a powerful method for interpreting gene
expression data by focusing on gene sets (pathways) rather than
individual genes. In the context of microbiome functional prediction,
GSEA can help identify pathways that are enriched in different
conditions.</p>
</div>
<div id="installation" class="section level2">
<h2>Installation</h2>
<p>First, make sure you have the necessary packages installed:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="co"># Install ggpicrust2</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;ggpicrust2&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>)) {</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a>  devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;cafferychen777/ggpicrust2&quot;</span>)</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a>}</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="co"># Install required Bioconductor packages</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>)) {</span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a>}</span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="fu">c</span>(<span class="st">&quot;fgsea&quot;</span>, <span class="st">&quot;clusterProfiler&quot;</span>, <span class="st">&quot;enrichplot&quot;</span>, <span class="st">&quot;DOSE&quot;</span>, <span class="st">&quot;pathview&quot;</span>))</span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a></span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a><span class="co"># Load the package</span></span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a><span class="fu">library</span>(ggpicrust2)</span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span></code></pre></div>
</div>
<div id="basic-gsea-analysis" class="section level2">
<h2>Basic GSEA Analysis</h2>
<p>Let’s start with a basic GSEA analysis using the example data
provided in the ggpicrust2 package:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># Load example data</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="fu">data</span>(ko_abundance)</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="fu">data</span>(metadata)</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="co"># Prepare abundance data</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>abundance_data <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(ko_abundance)</span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="fu">rownames</span>(abundance_data) <span class="ot">&lt;-</span> abundance_data[, <span class="st">&quot;#NAME&quot;</span>]</span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a>abundance_data <span class="ot">&lt;-</span> abundance_data[, <span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="co"># Run GSEA analysis</span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a>gsea_results <span class="ot">&lt;-</span> <span class="fu">pathway_gsea</span>(</span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a>  <span class="at">abundance =</span> abundance_data,</span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a>  <span class="at">metadata =</span> metadata,</span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a>  <span class="at">group =</span> <span class="st">&quot;Environment&quot;</span>,</span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a>  <span class="at">pathway_type =</span> <span class="st">&quot;KEGG&quot;</span>,</span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a>  <span class="at">method =</span> <span class="st">&quot;fgsea&quot;</span>,</span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a>  <span class="at">rank_method =</span> <span class="st">&quot;signal2noise&quot;</span>,</span>
<span id="cb2-18"><a href="#cb2-18" tabindex="-1"></a>  <span class="at">nperm =</span> <span class="dv">1000</span>,</span>
<span id="cb2-19"><a href="#cb2-19" tabindex="-1"></a>  <span class="at">min_size =</span> <span class="dv">10</span>,</span>
<span id="cb2-20"><a href="#cb2-20" tabindex="-1"></a>  <span class="at">max_size =</span> <span class="dv">500</span>,</span>
<span id="cb2-21"><a href="#cb2-21" tabindex="-1"></a>  <span class="at">p.adjust =</span> <span class="st">&quot;BH&quot;</span>,</span>
<span id="cb2-22"><a href="#cb2-22" tabindex="-1"></a>  <span class="at">seed =</span> <span class="dv">42</span></span>
<span id="cb2-23"><a href="#cb2-23" tabindex="-1"></a>)</span>
<span id="cb2-24"><a href="#cb2-24" tabindex="-1"></a></span>
<span id="cb2-25"><a href="#cb2-25" tabindex="-1"></a><span class="co"># View the top results</span></span>
<span id="cb2-26"><a href="#cb2-26" tabindex="-1"></a><span class="fu">head</span>(gsea_results)</span></code></pre></div>
</div>
<div id="annotating-gsea-results" class="section level2">
<h2>Annotating GSEA Results</h2>
<p>To make the results more interpretable, we can annotate them with
pathway names and descriptions:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># Annotate GSEA results</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>annotated_results <span class="ot">&lt;-</span> <span class="fu">gsea_pathway_annotation</span>(</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>  <span class="at">gsea_results =</span> gsea_results,</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>  <span class="at">pathway_type =</span> <span class="st">&quot;KEGG&quot;</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>)</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="co"># View the annotated results</span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a><span class="fu">head</span>(annotated_results)</span></code></pre></div>
</div>
<div id="visualizing-gsea-results" class="section level2">
<h2>Visualizing GSEA Results</h2>
<p>The ggpicrust2 package provides several visualization options for
GSEA results:</p>
<div id="barplot" class="section level3">
<h3>Barplot</h3>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co"># Create a barplot of the top enriched pathways</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>barplot <span class="ot">&lt;-</span> <span class="fu">visualize_gsea</span>(</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>  <span class="at">gsea_results =</span> annotated_results,</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>  <span class="at">plot_type =</span> <span class="st">&quot;barplot&quot;</span>,</span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>  <span class="at">n_pathways =</span> <span class="dv">20</span>,</span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>  <span class="at">sort_by =</span> <span class="st">&quot;p.adjust&quot;</span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>)</span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a><span class="co"># Display the plot</span></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a>barplot</span></code></pre></div>
</div>
<div id="dotplot" class="section level3">
<h3>Dotplot</h3>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># Create a dotplot of the top enriched pathways</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>dotplot <span class="ot">&lt;-</span> <span class="fu">visualize_gsea</span>(</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>  <span class="at">gsea_results =</span> annotated_results,</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>  <span class="at">plot_type =</span> <span class="st">&quot;dotplot&quot;</span>,</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a>  <span class="at">n_pathways =</span> <span class="dv">20</span>,</span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a>  <span class="at">sort_by =</span> <span class="st">&quot;p.adjust&quot;</span></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a>)</span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="co"># Display the plot</span></span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a>dotplot</span></code></pre></div>
</div>
<div id="enrichment-plot" class="section level3">
<h3>Enrichment Plot</h3>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># Create an enrichment plot for a specific pathway</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>enrichment_plot <span class="ot">&lt;-</span> <span class="fu">visualize_gsea</span>(</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>  <span class="at">gsea_results =</span> annotated_results,</span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>  <span class="at">plot_type =</span> <span class="st">&quot;enrichment_plot&quot;</span>,</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>  <span class="at">n_pathways =</span> <span class="dv">10</span>,</span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>  <span class="at">sort_by =</span> <span class="st">&quot;NES&quot;</span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a>)</span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a></span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a><span class="co"># Display the plot</span></span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a>enrichment_plot</span></code></pre></div>
</div>
</div>
<div id="comparing-gsea-and-daa-results" class="section level2">
<h2>Comparing GSEA and DAA Results</h2>
<p>It can be informative to compare the results from GSEA with those
from Differential Abundance Analysis (DAA):</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="co"># Run DAA analysis</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>daa_results <span class="ot">&lt;-</span> <span class="fu">pathway_daa</span>(</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>  <span class="at">abundance =</span> abundance_data,</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>  <span class="at">metadata =</span> metadata,</span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>  <span class="at">group =</span> <span class="st">&quot;Environment&quot;</span>,</span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a>  <span class="at">daa_method =</span> <span class="st">&quot;ALDEx2&quot;</span></span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a>)</span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a></span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a><span class="co"># Annotate DAA results</span></span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a>annotated_daa_results <span class="ot">&lt;-</span> <span class="fu">pathway_annotation</span>(</span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a>  <span class="at">pathway =</span> <span class="st">&quot;KO&quot;</span>,</span>
<span id="cb7-12"><a href="#cb7-12" tabindex="-1"></a>  <span class="at">daa_results_df =</span> daa_results,</span>
<span id="cb7-13"><a href="#cb7-13" tabindex="-1"></a>  <span class="at">ko_to_kegg =</span> <span class="cn">TRUE</span></span>
<span id="cb7-14"><a href="#cb7-14" tabindex="-1"></a>)</span>
<span id="cb7-15"><a href="#cb7-15" tabindex="-1"></a></span>
<span id="cb7-16"><a href="#cb7-16" tabindex="-1"></a><span class="co"># Compare GSEA and DAA results</span></span>
<span id="cb7-17"><a href="#cb7-17" tabindex="-1"></a>comparison <span class="ot">&lt;-</span> <span class="fu">compare_gsea_daa</span>(</span>
<span id="cb7-18"><a href="#cb7-18" tabindex="-1"></a>  <span class="at">gsea_results =</span> annotated_results,</span>
<span id="cb7-19"><a href="#cb7-19" tabindex="-1"></a>  <span class="at">daa_results =</span> annotated_daa_results,</span>
<span id="cb7-20"><a href="#cb7-20" tabindex="-1"></a>  <span class="at">plot_type =</span> <span class="st">&quot;venn&quot;</span>,</span>
<span id="cb7-21"><a href="#cb7-21" tabindex="-1"></a>  <span class="at">p_threshold =</span> <span class="fl">0.05</span></span>
<span id="cb7-22"><a href="#cb7-22" tabindex="-1"></a>)</span>
<span id="cb7-23"><a href="#cb7-23" tabindex="-1"></a></span>
<span id="cb7-24"><a href="#cb7-24" tabindex="-1"></a><span class="co"># Display the comparison plot</span></span>
<span id="cb7-25"><a href="#cb7-25" tabindex="-1"></a>comparison<span class="sc">$</span>plot</span>
<span id="cb7-26"><a href="#cb7-26" tabindex="-1"></a></span>
<span id="cb7-27"><a href="#cb7-27" tabindex="-1"></a><span class="co"># View the comparison results</span></span>
<span id="cb7-28"><a href="#cb7-28" tabindex="-1"></a>comparison<span class="sc">$</span>results</span></code></pre></div>
</div>
<div id="integrated-analysis-with-ggpicrust2_extended" class="section level2">
<h2>Integrated Analysis with ggpicrust2_extended</h2>
<p>For a more streamlined workflow, you can use the
<code>ggpicrust2_extended</code> function, which integrates both DAA and
GSEA analyses:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co"># Run integrated analysis</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>integrated_results <span class="ot">&lt;-</span> <span class="fu">ggpicrust2_extended</span>(</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>  <span class="at">data =</span> ko_abundance,</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>  <span class="at">metadata =</span> metadata,</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a>  <span class="at">group =</span> <span class="st">&quot;Environment&quot;</span>,</span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a>  <span class="at">pathway =</span> <span class="st">&quot;KO&quot;</span>,</span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a>  <span class="at">daa_method =</span> <span class="st">&quot;LinDA&quot;</span>,</span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a>  <span class="at">ko_to_kegg =</span> <span class="cn">TRUE</span>,</span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a>  <span class="at">run_gsea =</span> <span class="cn">TRUE</span>,</span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a>  <span class="at">gsea_params =</span> <span class="fu">list</span>(</span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a>    <span class="at">method =</span> <span class="st">&quot;fgsea&quot;</span>,</span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a>    <span class="at">rank_method =</span> <span class="st">&quot;signal2noise&quot;</span>,</span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a>    <span class="at">nperm =</span> <span class="dv">1000</span></span>
<span id="cb8-14"><a href="#cb8-14" tabindex="-1"></a>  )</span>
<span id="cb8-15"><a href="#cb8-15" tabindex="-1"></a>)</span>
<span id="cb8-16"><a href="#cb8-16" tabindex="-1"></a></span>
<span id="cb8-17"><a href="#cb8-17" tabindex="-1"></a><span class="co"># Access DAA results</span></span>
<span id="cb8-18"><a href="#cb8-18" tabindex="-1"></a>daa_results <span class="ot">&lt;-</span> integrated_results<span class="sc">$</span>daa_results</span>
<span id="cb8-19"><a href="#cb8-19" tabindex="-1"></a></span>
<span id="cb8-20"><a href="#cb8-20" tabindex="-1"></a><span class="co"># Access GSEA results</span></span>
<span id="cb8-21"><a href="#cb8-21" tabindex="-1"></a>gsea_results <span class="ot">&lt;-</span> integrated_results<span class="sc">$</span>gsea_results</span>
<span id="cb8-22"><a href="#cb8-22" tabindex="-1"></a></span>
<span id="cb8-23"><a href="#cb8-23" tabindex="-1"></a><span class="co"># Access plots</span></span>
<span id="cb8-24"><a href="#cb8-24" tabindex="-1"></a>daa_plot <span class="ot">&lt;-</span> integrated_results<span class="sc">$</span>daa_plot</span>
<span id="cb8-25"><a href="#cb8-25" tabindex="-1"></a>gsea_plot <span class="ot">&lt;-</span> integrated_results<span class="sc">$</span>gsea_plot</span></code></pre></div>
</div>
<div id="conclusion" class="section level2">
<h2>Conclusion</h2>
<p>Gene Set Enrichment Analysis provides a complementary approach to
Differential Abundance Analysis for interpreting PICRUSt2 predicted
functional data. By focusing on pathways rather than individual
features, GSEA can help identify biologically meaningful patterns that
might be missed by traditional methods.</p>
<p>The ggpicrust2 package now offers a comprehensive suite of tools for
both DAA and GSEA analyses, making it easier to gain insights from
microbiome functional prediction data.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
